yjit/src/core.rs

```diff
@@ -1776,55 +1789,42 @@ pub fn defer_compilation(jit: &JITState, cur_ctx: &Context, cb: &mut CodeBlock,
 // Remove all references to a block then free it.
 fn free_block(blockref: &BlockRef)
 {
-    todo!("free_block unimplemented");
-    /*
-    yjit_unlink_method_lookup_dependency(block);
-    yjit_block_assumptions_free(block);
+    use crate::invariants::*;
+
+    block_assumptions_free(blockref);
+
+    let block = blockref.borrow();
 
     // Remove this block from the predecessor's targets
-    rb_darray_for(block->incoming, incoming_idx) {
+    for pred_branchref in &block.incoming {
         // Branch from the predecessor to us
-        branch_t *pred_branch = rb_darray_get(block->incoming, incoming_idx);
+        let mut pred_branch = pred_branchref.borrow_mut();
 
         // If this is us, nullify the target block
-        for (size_t succ_idx = 0; succ_idx < 2; succ_idx++) {
-            if (pred_branch->blocks[succ_idx] == block) {
-                pred_branch->blocks[succ_idx] = NULL;
+        for pred_succ_ref in &mut pred_branch.blocks {
+            if let Some(pred_succ) = pred_succ_ref {
+                if pred_succ == blockref {
+                    *pred_succ_ref = None;
+                }
             }
         }
     }
 
     // For each outgoing branch
-    rb_darray_for(block->outgoing, branch_idx) {
-        branch_t *out_branch = rb_darray_get(block->outgoing, branch_idx);
+    for out_branchref in &block.outgoing {
+        let out_branch = out_branchref.borrow();
 
         // For each successor block
-        for (size_t succ_idx = 0; succ_idx < 2; succ_idx++) {
-            block_t *succ = out_branch->blocks[succ_idx];
-
-            if (succ == NULL)
-                continue;
-
-            // Remove this block from the successor's incoming list
-            rb_darray_for(succ->incoming, incoming_idx) {
-                branch_t *pred_branch = rb_darray_get(succ->incoming, incoming_idx);
-                if (pred_branch == out_branch) {
-                    rb_darray_remove_unordered(succ->incoming, incoming_idx);
-                    break;
-                }
+        for succ in &out_branch.blocks {
+            if let Some(succ) = succ {
+                // Remove outgoing branch from the successor's incoming list
+                let mut succ_block = succ.borrow_mut();
+                succ_block.incoming.retain(|succ_incoming| !Rc::ptr_eq(succ_incoming, out_branchref));
             }
         }
-
-        // Free the outgoing branch entry
-        free(out_branch);
     }
 
-    rb_darray_free(block->incoming);
-    rb_darray_free(block->outgoing);
-    rb_darray_free(block->gc_object_offsets);
```

Comment:
Should we be clearing these arrays here (setting their length to zero)? Just to drop the references and indicate that this block has been freed.

Reply:
Hmm in the event that we have lingering reference to the block which prevents deallocation having these info might be useful for debugging. Maybe we should add a debug only invalidated field though.